### ISSUE TITLE
Build Config - modules exclude option - possible bug?

### ISSUE THREAD

**Issue by kuakman** [requirejs/r.js#936]
Hello,

First of all, I wanted to apologies before hand if this doesn't end up being a bug. My intention is not to pollute the issue's list at all, but this behavior caught my attention in multiple occasions and the reason why I decided to post it here (instead of stackoverflow) is that it may become a possible feature if I didn't miss something from the documentation. I still think this may be a case of "expected behavior" though but I think is worth a shot and share this case with the author, maintainers or people around here.

Information:
- Running on a Mac OSX: 10.10.5
- RequireJS version: 2.3.0
- Node Version: 6.2.2

This is a [gist](https://gist.github.com/kuakman/eda29528776d437b877e58ebb3963d93) link with information I will be referencing through the explanation of my case.

Here it goes:

I’m trying to automate the build process using requireJS optimizer, so developers at my company they shouldn’t need to touch the build profile.
The goal was to provide a way for developers to control how different files should be ‘bundled’, inside the code base rather than editing the build config.

To do that, I have a small script that generates the ‘modules’ property automatically by traversing a folder structure recursively and using the information it gets to build the module entries.
Consider the following convention:
(Notice that the specific location of fileA-1.js, fileA-2.js or fileB-1.js doesn’t actually matter, this is just to illustrate the case in a simple way).

<pre>
+ src
  |
  + folderA
  |    |
  |    + subfolderA2
  |    |     |
  |    |     | > fileA2-1.js
  |    |     | > moduleA2-bundle.js
  |    |   
  |    | > fileA-1.js
  |    | > fileA-2.js
  |    | > moduleA-bundle.js
  |    
  + folderB
  |    |
  |    | > fileB-1.js
  |    | > moduleB-bundle.js
  |
  ...
</pre>

Before r.js is fired, the script will pull **only** the files ending in **-bundle.js** in order to define the **exclude** rules.
And this is how the “\*-bundle.js” files look like internally (notice that I'm using babel and amd transformer to convert the **imports** to **define([], function() { })** in a previous step. That works just fine, not a concern.

**moduleA2-bundle.js:**
<pre>
import 'folderA/subfolderA2/fileA2-1';
</pre>
**moduleA-bundle.js**
<pre>
import 'folderA/fileA-1';
import 'folderA/fileA-2';
</pre>
**moduleB-bundle.js**
<pre>
import 'folderB/fileB-1';
</pre>

So, the script will build up the following automatically before firing r.js.

<pre>
…
modules: [{
    name: ‘folderA/subfolderA2/moduleA2-bundle’,
    exclude: [‘folderA/moduleA-bundle’, ‘folderB/moduleB-bundle']
}, {
    name: ‘folderA/moduleA-bundle’,
    exclude: [‘folderA/subfolderA2/moduleA2-bundle’, ‘folderB/moduleB-bundle']
}, {
    name: ‘folderB/moduleB-bundle’,
    exclude: [‘folderA/subfolderA2/moduleA2-bundle’, ‘folderA/moduleA-bundle’]
}],
...
</pre>

You will notice that this essentially iterates over each module and says:

*"exclude" everything else from the current module*.

You can see a real example in the gist link I provided.

So this will provide a quick way for developers to break down an app into multiple pieces and have some level of control how things should be bundled via these **'*-bundle.js'** files. Instead of touching a monolithic build profile and dealing with ‘includes’, ‘exclude’, ‘excludeShallow’, adding new module entries and so on.

Now, if you look at the gist link, you will notice there is something wrong with one of the bundles in build.txt output sample:

<pre>
...
store/session/store-session-bundle.js
----------------
** They should be here... **
store/session/store-session-bundle.js

ui/common/ui-common-bundle.js
----------------
ui/common/nav/nav.js
ui/common/toolbar/omcnavbar.js
store/session/actions/session-actions.js
store/session/reducers/session-reducer.js
store/session/session-store.js
^ They shouldn't be here...
ui/common/toolbar/toolbar.js
ui/common/layout/header/header.js
ui/common/breadcrumbs/breadcrumbs.js
ui/common/layout/footer/footer.js
ui/common/layout/page.js
ui/common/layout/support.js
ui/common/tabs/tabs.js
ui/common/ui-common-bundle.js
...
</pre>

And here is the issue that again, it might be "expected behavior". Even if you tell the optimizer to exclude everything else for "ui/common/ui-common-bundle.js” (like **store/session/store-session-bundle.js**), it still bundles the store/session/**/*.js together inside ui-common-bundle.js.

So after seeing this behavior, I'm assuming (and I’m not sure if this is true) that requirejs internally calculates the dependencies based on the 'order' or how many times the dependencies are being required, to decide where to put them.

The second assumption I was trying to prove using this strategy:

The **exclude** rule will **force** to exclude everything else "no matter what" and only include the dependencies declared inside a bundle, independent of how many times (for example) a file inside **ui-common-bundle.js** requires a file located in a different targeted bundle, like in this case: **store-session-bundle.js**.

If that’s the case, it’s there a way to enforce it? Maybe a feature to make this happen?

I'm afraid that there must be a different way to achieve this or I'm a little confuse on the usage of the "exclude" option. Not familiar with the requirejs code base yet.
I also took a look at the doc notes attached to the example build file and the second sentence caught my attention:

<pre>
...
//This module entry combines all the dependencies of foo/bar/bip into one file,
//but excludes foo/bar/bop and its dependencies from the built file. >>> If you want
//to exclude a module that is also another module being optimized, it is more
//efficient if you define that module optimization entry before using it
//in an exclude array. <<<<
{
   name: "foo/bar/bip",
   exclude: [
       "foo/bar/bop"
   ]
},
...
</pre>

Then the rest of the process is including the bundles directive in the final config.js, and I wanted to thank you @jrburke for adding the bundleConfigOutFile property feature.
Before I was doing the same thing but manually. Now magically it gets that directive injected automatically ;).

Thanks again and sorry if my english is not great. Any feedback will be appreciate it.

**Comment by kuakman** [comment]
After publishing this issue I noticed there are a few guys with the same question. They called globalexclude which is a similar case #266. I have seen one comment that @jrburke posted in that issue that is helping me to understand the exclude option:

> When a previous layer name is listed as an exclude for another layer, that previous layer's modules are excluded from that layer.

So I might need to grasp the idea by prototyping the algorithm following that rule:

Once you define a exclude list, the script will need to iterate over all the previous ones of all previous module entries to wether, include that layer or not into the current exclude list.
It might get a little bit dirty and confusing but worth to try it out.

**Comment by jrburke** [comment]
I am not sure I follow exactly, but it sounds like the rules in r.js are not meeting your needs. The rules that r.js uses, using this sample config to help explain it:

```javascript
modules : [
{ name: 'a', exclude: ['c'] }
{ name: 'b', exclude: ['a'] }
]
```

All modules for 'a' are considered part of the 'a' `modules` built output. So if 'a' depended on 'c', 'd', and 'e', then 'c', 'd' and 'e' would end up in the 'a' bundle. However, with the excludes, it means the 'a' bundle will contain 'd', 'e', 'a'. That is considered the 'a' bundle, and that is the module set used for any other `modules` entries defined below it that mention excluding 'a'.

So when 'b' is traced, and say it depends on 'f' and 'e' and 'c', it will just end up with 'f', 'c', and 'b': since the 'a' bundle has 'e' in it, that is excluded from the 'b' bundle and even though 'a' depends on 'c', since it was excluded from the 'a' bundle, then it is not excluded from 'b'.

And that exclude behavior only works for `modules` entries that are set up before the `modules` entry that needs it.

At least that is how I recall it working, it has been a while since I have specifically traced it. 

There are no plans to change that behavior, but if you would like to do some fancier bundling/exclusions, I suggest using [amodro-trace](https://github.com/amodrojs/amodro-trace) instead. It just does the tracing of dependencies and you can decides how to partition those traces into the bundles you need.

Closing as a discussion ticket, but feel free to continue discussion here.

**Comment by kuakman** [comment]
@jrburke Thanks for your quick response and your clarification. I really appreciate it.

But yea, based on your explanation, the order of those module entries affects the end result.
I actually created this [repo](https://github.com/kuakman/require-experiment) to put your example to work and it makes sense.

But unfortunately trying to workaround it, to make it work the way I want seems a little bit tricky.
So, to automate mutual exclusion of [N] number of modules despite their order, I'm going to need to customize the implementation. **amodro-trace** seems to be the answer to my question. I started looking at it and it's pretty neat. Nice job!
